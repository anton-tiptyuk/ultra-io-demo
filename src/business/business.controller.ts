import { Controller, Get } from '@nestjs/common';

import { BusinessService } from './business.service';

@Controller('business')
export class BusinessController {
  constructor(private readonly service: BusinessService) { }

  @Get('trigger-aging')
  async triggerAging() {
    await this.service.performAging();
    return 'aging performed';
  }
}
