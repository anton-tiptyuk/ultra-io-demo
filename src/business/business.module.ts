import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';

import { Manufacturer, Owner, CarToOwner } from '../db/models';

import { BusinessController } from './business.controller';
import { BusinessService } from './business.service';

@Module({
  imports: [TypeOrmModule.forFeature([Manufacturer, Owner, CarToOwner])],
  controllers: [BusinessController],
  providers: [BusinessService],
})
export class BusinessModule { }
