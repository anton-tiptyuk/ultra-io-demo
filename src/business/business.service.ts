import { Repository } from 'typeorm';
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';

import { Owner } from '../db/models';
import { CarService } from '../car/car.service';

@Injectable()
export class BusinessService {
  constructor(
    private readonly carService: CarService,
    @InjectRepository(Owner)
    private readonly ownerRepo: Repository<Owner>
  ) { }

  // First we call carService
  // It applies the car price with discount
  // and drops CarToOwner join table entitites with purchaseDate matching range

  // Then we get back here and drop orphan owners
  // (because even though we seed data in a way owner posesses only one car,
  // the data structure and common logic says it is possible that owner had owned several cars
  // in the past)
  async performAging() {
    const results: { affected?: number | null }[] = await this.carService.performAging();

    results.push(
      await this
        .ownerRepo
        .createQueryBuilder()
        .delete()
        .where('not exists (select 1 from car_to_owner where car_to_owner.owner_id = owner.id)')
        .execute()
    );

    return results.map(({ affected }) => affected);
  }

}
