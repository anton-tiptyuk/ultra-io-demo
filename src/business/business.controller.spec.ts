import { Test, TestingModule } from '@nestjs/testing';
import { BusinessController } from './business.controller';
import { BusinessService } from './business.service';

let controller: BusinessController;

const serviceFake = { performAging: () => undefined };
const performAgingSpy = jest.spyOn(serviceFake, 'performAging');

describe('Business Controller', () => {
  beforeAll(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [{
        provide: BusinessService,
        useValue: serviceFake,
      }],
      controllers: [BusinessController],
    }).compile();

    controller = module.get<BusinessController>(BusinessController);
  });

  it('should be defined', () => {
    expect(controller).toBeDefined();
  });

  it('should trigger car and owner aging', async () => {
    performAgingSpy.mockImplementation(async () => 'any kind of result service returns');
    await expect(controller.triggerAging()).resolves.toMatchSnapshot();
  });
});
