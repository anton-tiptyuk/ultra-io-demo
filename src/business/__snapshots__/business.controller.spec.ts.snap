// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Business Controller should trigger car and owner aging 1`] = `
Object {
  "entitiesAffected": "any kind of result service returns",
  "success": true,
}
`;
